File: deploy/cloudflare/workers/shared/memory-privacy-receipts.ts
/** Same HMAC domain as the upstream canonical privacy deletion receipt. */
export async function memoryPrivacyReceiptId(
  secret: string | undefined,
  uid: string,
  memoryId: string,
): Promise<string> {
  const encoder = new TextEncoder();
  if (typeof secret !== "string" || encoder.encode(secret).length < 32)
    throw new Error("memory privacy receipt secret is unavailable");
  if (!uid || !memoryId || uid.includes("\n") || memoryId.includes("\n"))
    throw new Error("invalid memory privacy identity");
  const key = await crypto.subtle.importKey(
    "raw",
    encoder.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const digest = await crypto.subtle.sign(
    "HMAC",
    key,
    encoder.encode(`memory-privacy-receipt.v2\n${uid}\n${memoryId}`),
  );
  return (
    "receipt_" +
    Array.from(new Uint8Array(digest), (byte) =>
      byte.toString(16).padStart(2, "0"),
    ).join("")
  );
}

File: deploy/cloudflare/python/api-core/src/memory_privacy_receipts.py
"""Opaque deletion identities shared by every memory creator and privacy apply.

The HMAC domain and output match upstream privacy_deletion_receipt_id. A
dedicated deployment secret keeps receipt lifetime independent of auth rotation.
The D1 receipt gate owns the final, transaction-time admission decision.
"""

import hashlib
import hmac


def privacy_receipt_id(env, uid: str, memory_id: str) -> str:
    secret = getattr(env, 'MEMORY_PRIVACY_SECRET', None)
    if not isinstance(secret, str) or len(secret.encode('utf-8')) < 32:
        raise ValueError('memory privacy receipt secret is unavailable')
    if not uid or not memory_id or '\n' in uid or '\n' in memory_id:
        raise ValueError('invalid memory privacy identity')
    digest = hmac.new(
        secret.encode('utf-8'),
        f'memory-privacy-receipt.v2\n{uid}\n{memory_id}'.encode('utf-8'),
        hashlib.sha256,
    ).hexdigest()
    return f'receipt_{digest}'
